euphoria-spark/src/main/java/cz/seznam/euphoria/spark/SortTranslator.java

```diff
@@ -0,0 +1,156 @@
+  @Override
+  @SuppressWarnings("unchecked")
+  public JavaRDD<?> translate(Sort operator,
+                              SparkExecutorContext context) {
+
+    final JavaRDD<SparkElement> input = (JavaRDD) context.getSingleInput(operator);
+    
+    final UnaryFunction<Object, Integer> keyExtractor = operator.getKeyExtractor();
+    final UnaryFunction<Object, Comparable> sortByFn = operator.getSortByExtractor();
+    final ExtractEventTime eventTimeAssigner = operator.getEventTimeAssigner();
+    final Windowing windowing = operator.getWindowing() == null
+            ? AttachedWindowing.INSTANCE
+            : operator.getWindowing();
+
+    // ~ extract key/value + timestamp from input elements and assign windows
+    JavaPairRDD<Tuple3<Integer, Window, Comparable>, Object> tuples = input.flatMapToPair(
+            new CompositeKeyExtractor(keyExtractor, sortByFn, windowing, eventTimeAssigner));
+    
+    Partitioner partitioner = new PartitioningWrapper(operator.getPartitioning().getNumPartitions());
+    Comparator comparator = new TripleComparator();
+
+    JavaPairRDD<Tuple3<Integer, Window, Comparable>, Object> sorted = 
+        tuples.repartitionAndSortWithinPartitions(partitioner, comparator);
+    
+    return sorted.values();
+  }
+  
+  /**
+   * Extracts {@link Window} from {@link SparkElement} and assigns timestamp 
+   * according to (optional) eventTimeAssigner.
+   * The result composite key tuple consists of [partitionId, window, sortByKey].
+   */
+  private static class CompositeKeyExtractor
+          implements PairFlatMapFunction<SparkElement, Tuple3<Integer, Window, Comparable>, Object> {
+
+    private final UnaryFunction<Object, Integer> keyExtractor;
+    private final UnaryFunction<Object, Comparable> sortByFn;
+    private final Windowing windowing;
+    @Nullable
+    private final ExtractEventTime eventTimeAssigner;
+
+    public CompositeKeyExtractor(UnaryFunction<Object, Integer> keyExtractor,
+                                 UnaryFunction<Object, Comparable> sortByFn,
+                                 Windowing windowing,
+                                 @Nullable ExtractEventTime eventTimeAssigner) {
+      this.keyExtractor = keyExtractor;
+      this.sortByFn = sortByFn;
+      this.windowing = windowing;
+      this.eventTimeAssigner = eventTimeAssigner;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Iterator<Tuple2<Tuple3<Integer, Window, Comparable>, Object>> call(SparkElement wel) throws Exception {
+      if (eventTimeAssigner != null) {
+        wel.setTimestamp(eventTimeAssigner.extractTimestamp(wel.getElement()));
+      }
+
+      Iterable<Window> windows = windowing.assignWindowsToElement(wel);
+      List<Tuple2<Tuple3<Integer, Window, Comparable>, Object>> out = new ArrayList<>();
+      for (Window wid : windows) {
+        Object el = wel.getElement();
+        out.add(new Tuple2<>(
+            new Tuple3(keyExtractor.apply(el), wid, sortByFn.apply(el)), wel));
```

Comment:
I am afraid it's wrong here. Output element (`wel`) carries the original window, not the newly assigned. It seems this situation can be solved the same way as in `ReduceByKeyTranslator` by extracting `TimestampedElement` and reconstructing `SparkElement` in the final map operation.
Also it shows windowing is not properly covered in operator-test-kit.